### ISSUE TITLE
Expose an API for manually triggering prefetching for a pathname

### ISSUE THREAD

**Issue by schalkventer** [gatsbyjs/gatsby#8122]
## Summary

Currently there is no way to programmatically trigger the prefetching  done for a pathname when the `<Link>` component is mounted.

### Basic example

```
const fireEvent = () => {
  const id = createNewPerson();
  navigate(`/person?id=${id}`);
}
```

```
<button onClick={fireEvent}>Create Person</button>
```

### Motivation

Developers might want to leverage prefetching when programmatically triggering `navigate`.

**Comment by kakadiadarpan** [comment]
@schalkventer Currently, if anyone hovers over a `Link` component, the resources required for that path are prefetched by default.

@pieh @m-allanson @KyleAMathews any suggestions to leverage prefetching for programmatic navigation?

**Comment by KyleAMathews** [comment]
This seems pretty reasonable to me. 

The API would be something like:

```javascript
import { prefetchPathname } from "gatsby"

prefetchPathname(`/about/`)
```

**Comment by schalkventer** [comment]
@KyleAMathews, if it is as straightforward as the above it seems like a good solution. 👍 

For the sake of clarity, what is the baseline behaviour for preloading via `Link`? Is it as indicated above only once you hover over the link, or is it on the rendering of the view that contains the link?

**Comment by KyleAMathews** [comment]
It's when the link appears in the viewport. We leverage https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API for this.

**Comment by schalkventer** [comment]
Has this been implemented in Gatsby 2? I.e. can it be closed?

**Comment by devrchancay** [comment]
@schalkventer I would think that if it can be closed.

**Comment by KyleAMathews** [comment]
Would love a PR from you for this? Is this something you'd like to work on in the near future? I can provide some hints on how to do it.

**Comment by schalkventer** [comment]
@KyleAMathews For sure!

Do you have any suggestions where I can start looking to find this mechanism in the code?

**Comment by KyleAMathews** [comment]
Great!

The entry module for the browser for `gatsby` is https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/gatsby-browser-entry.js so that's where `prefetchPathname` should be exported from.

The function which does the prefetching is `enqueue` in `loader.js` https://github.com/gatsbyjs/gatsby/blob/0124389f81df542cd0c22fa68fbe980cb04b0120/packages/gatsby/cache-dir/loader.js#L223

Then you can just import that directly into gatsby-browser-entry.js and then reexport it as `prefetchPathname`

**Comment by schalkventer** [comment]
Awesome! Thank @KyleAMathews 🙏.

Will have a look!

**Comment by schalkventer** [comment]
@KyleAMathews Can you re-open this issue and assign it to me?

**Comment by KyleAMathews** [comment]
Done!

**Comment by thelovesmith** [comment]
Hello good morning! 
I keep getting this warning in development due to 'prefetchPathname":
`
 warning  in ./.cache/gatsby-browser-entry.js
⠀
"export 'enqueue' (reexported as 'prefetchPathname') was not found in './loader'
`

Wanted to know what is going on here and any solutions to get rid of this warning?

**Comment by wardpeet** [comment]
we got this fixed in gatsby version 2.13.24

**Comment by thelovesmith** [comment]
hmm I appreciate you responding so  quickly, but i literally just updated to version 2.13.25 and my the dev server just crashed. lol def was working before. I know now this is a separate issue lol Ill try and find a solution elsewhere thanks 
`
There was a problem loading the local develop command. Gatsby may not be installed. Perhaps you need to run "npm install"? Cannot find module 'gatsby-cli/lib/reporter'
Require stack:
- /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/redux/actions/public.js
- /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/redux/actions/index.js
- /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/utils/webpack.config.js
- /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/commands/develop.js
- /usr/local/lib/node_modules/gatsby-cli/lib/create-cli.js
- /usr/local/lib/node_modules/gatsby-cli/lib/index.js



  Error: Cannot find module 'gatsby-cli/lib/reporter'
  Require stack:
  - /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/redux/actions/public.js
  - /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/redux/actions/index.js
  - /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/utils/webpack.config.js
  - /mnt/c/Users/PARFunding/Desktop/AHinds/gatsby-site/node_modules/gatsby/dist/commands/develop.js
  - /usr/local/lib/node_modules/gatsby-cli/lib/create-cli.js
  - /usr/local/lib/node_modules/gatsby-cli/lib/index.js
  
`

**Comment by wardpeet** [comment]
I would try reinstalling. Lookz like some deps are missing

**Comment by nleco** [comment]
was this ever implemented?

**Comment by nleco** [comment]
I am on gatsby 2.32.2 and can import `prefetchPathname`, however, when I use it It does not seem to do anything.
I don't see many discussion around this api nor documentation/caveats.

i tried via 
`useEffect(() => { prefetchPathname("/url") }, [])` as well as via  a **class:componentDidMount** method.

If I have a `Link` present, and hover over it, I see app-data.json and page-data.json load. 

I'm also logging the flow found in in the `gatsby/cache-dir/loader.js:prefetch()` function and it does go all the way and adds the path to the 'prefetchCompleted' Set.

I'm a bit confused.

**Comment by viggin543** [comment]
this does not work for me as well

**Comment by viggin543** [comment]
```typescript
export function usePreFetch() {
  const ref = useRef(null)
  useEffect(() => {
    // @ts-ignore
    ref?.current?.dispatchEvent(
      new MouseEvent('mouseover', {
        view: window,
        bubbles: true,
        cancelable: true
      })
    )
  }, [])

  return ref
}

export const PreFetch: FC<{ page: Page }> = ({ page }) => {
  const ref = usePreFetch()
  return <Link to={page} ref={ref} />
}
```

this solves the problem, an invisible link that trigers a synthetic hover event on page load which causes the pre fetch to work
